joomla/joomla-cms
# [login forms] performance improvements in two factor authentication

#### Description

This PR improves performance and memory usage of the admin login page, the offline page, all the pages that have a login module in the frontend and the component users view login form.

Joomla is loading fof just to check if any two factor auth plugins are enabled to show the "Secret key" field. IMHO we can use Joomla (already loaded) libraries for that.
This way we can have performance improvements(around 4-6ms PHP 5.6.16) in every page that use this login modules.
###### Performance before PR (frontend mod_login as example)

![image](https://cloud.githubusercontent.com/assets/9630530/12460488/d21a8b44-bfab-11e5-8da4-0bb994bfcad8.png)
###### Performance after PR (frontend mod_login as example)

![image](https://cloud.githubusercontent.com/assets/9630530/12496080/63d55782-c08b-11e5-914a-af66ee9dbb65.png)
#### How to test
1. Install Joomla with latest staging with sample data
2. Enable Google Authenticator (or Yubikey) two factor authentication
3. Apply patch 
4. Create a dummy user and configure it to use with Google Authenticator (if you don't have use a chrome plugin for that, for instance https://chrome.google.com/webstore/detail/authenticator/bhghoamapcdpbohphigoooaddinpkbai)
5. Check Two auth is working:
   - Go to frontend and try to login with the new user **without** the two auth secret key. You can't
   - Go to frontend and try to login with the new user **with** the two auth secret key. You can

Note: for test the performance go to any (or all this files):
- /modules/mod_login/mod_login.php
- /administrator/modules/mod_login/mod_login.php
- /templates/system/offline.php
- /components/com_users/views/login/view.html.php

And put a mark in the debug profiler before and after the call to get the two factor methods.
Example: /modules/mod_login/mod_login.php

``` php
JDEBUG ? JProfiler::getInstance('Application')->mark('- start getting twofactorauth methods') : null;
$twofactormethods = [....];
JDEBUG ? JProfiler::getInstance('Application')->mark('- <strong>Time to get</strong> twofactorauth methods') : null;
```
#### Observations

Suggestions, improvements and code reviews are welcome.
